File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool dfs(vector<vector<char>>& board, int row, int col,string& word,int index,vector<vector<bool>>& vis){
        if(index == word.size()-1){
            // cout << board[row][col] << " here" << endl;
            if(board[row][col] == word[index]) return true;
            return false;
        }
        
        int delrow[] = {+1,0,-1,0};
        int delcol[] = {0,+1,0,-1};

        for(int i=0;i<4;i++){
            int nrow = row + delrow[i];
            int ncol = col + delcol[i];
            if(nrow>=0 && nrow<board.size() && ncol>=0 && ncol<board[0].size() 
&& board[nrow][ncol] == word[index+1] && vis[nrow][ncol] == false){
                vis[nrow][ncol] = true;
                // cout <<board[row][col] << "  " << nrow << " " << ncol << endl;
                if(dfs(board,nrow,ncol,word,index+1,vis)==true) return true;
                vis[nrow][ncol] = false;
            }
        }
        
        return false;
    }
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size(), n = board[0].size();
        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[0].size();j++){
                if(board[i][j] == word[0]){
                    vector<vector<bool>> vis(m,vector<bool>(n,false));
                    vis[i][j] = true;
                    if(dfs(board,i,j,word,0,vis) == true) return true;
                }
                
            }
        }
        
        return false;
    }
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        
        def dfs(i, j, wordIdx):
            if wordIdx == len(word):  # Base case: full word matched
                return True

            # Direction vectors for right, down, left, up
            directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

            temp = board[i][j]
            board[i][j] = '#'  # Mark as visited
            
            for di, dj in directions:
                ni, nj = i + di, j + dj
                if 0 <= ni < m and 0 <= nj < n and board[ni][nj] == word[wordIdx]:
                    if dfs(ni, nj, wordIdx + 1):  # Recursively check next character
                        return True
            
            board[i][j] = temp  # Backtrack (undo visit)
            return False
        
        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0]:  # Start DFS if first letter matches
                    if dfs(i, j, 1):
                        return True
        
        return False